### PR TITLE
add spot instance support and vpc selection

### PR DESCRIPTION
**Pull request by benoitm974**
Still missing:
 - availability zone selection (default to a for docker-machine driver)
 - timezone list in cloudformation
 - spot instance price as cloudformation parameter

### CODE REVIEW

File: runner.yml [hunk 1]
```diff
@@ -60,7 +64,7 @@ Parameters:
     Type: 'String'
     Description: |
       Instance type for GitLab Runners.
-    AllowedValues: ['t2.nano', 't2.micro', 't2.small', 't2.medium', 't2.large', 't3.nano', 't3.micro', 't3.small', 't3.medium', 't3.large']
+    AllowedValues: ['t2.nano', 't2.micro', 't2.small', 't2.medium', 't2.large', 't3.nano', 't3.micro', 't3.small', 't3.medium', 't3.large', , 'm5.large', 'm5d.large']
```
Comment (fquffio) [inline runner.yml]: Maybe we should add all types of M5 and M5d instances? 🤔

File: runner.yml [hunk 2]
```diff
@@ -84,7 +88,7 @@ Parameters:
     Type: 'String'
     Description: |
       Timezone in which non-working hours are evaluated.
-    Default: 'Europe/Rome'
+    Default: 'Europe/Paris'
```
Comment (fquffio) [inline runner.yml]: Not a very relevant change, but I'd stick to the previous default. 😄 🇮🇹 🇫🇷 Or maybe we should use UTC. It's still configurable, after all.

File: runner.yml [hunk 3]
```diff
@@ -324,7 +331,7 @@ Resources:
                   executor = "docker+machine"
                   [runners.docker]
                     tls_verify = false
-                    image = "ubuntu:latest"
+                    image = "alpine"
```
Comment (fquffio) [inline runner.yml]: Probably it'd be better to add another parameter with the default image name, to make it configurable on a per-stack basis.

File: runner.yml [hunk 4]
```diff
@@ -338,12 +345,12 @@ Resources:
                   [runners.machine]
                     IdleCount = ${GitlabIdleCount}
                     IdleTime = ${GitlabIdleTime}
-                    MaxBuilds = ${GitlabMaxConcurrentBuilds}
+                    MaxBuilds = 100
```
Comment (fquffio) [inline runner.yml]: I'd revert this change, and use the `GitlabMaxConcurrentBuilds` parameter. You may still set it to 100 in your stack.

File: runner.yml [hunk 5]
```diff
@@ -338,12 +345,12 @@ Resources:
...
                     MachineDriver = "amazonec2"
                     MachineName = "gitlab-docker-machine-%s"
-                    MachineOptions = ["amazonec2-instance-type=${GitlabRunnerInstanceType}", "amazonec2-region=${AWS::Region}", "amazonec2-vpc-id=${VpcId}", "amazonec2-security-group=${AWS::StackName}-RunnersSecurityGroup", "amazonec2-use-private-address=true", "amazonec2-iam-instance-profile=${RunnersInstanceProfile}"]
+                    MachineOptions = ["engine-storage-driver=overlay2", "amazonec2-instance-type=${GitlabRunnerInstanceType}", "amazonec2-region=${AWS::Region}", "amazonec2-vpc-id=${VpcId}", "amazonec2-security-group=${AWS::StackName}-RunnersSecurityGroup", "amazonec2-use-private-address=true", "amazonec2-iam-instance-profile=${RunnersInstanceProfile}", "amazonec2-request-spot-instance=true", "amazonec2-spot-price=0.08"]
```
Comment (fquffio) [inline runner.yml]: * `"engine-storage-driver=overlay2"` this is surely an improvement. I wonder if we should make this configurable, but it's probably good so.
* `"amazonec2-request-spot-instance=true", "amazonec2-spot-price=0.08"`: I don't like the fact that spot instances are used in any case. Somebody may still want to use on-demand instances. Also, this hardcoded bid price is good for `m5.large` instances, but it's definitely too much for, say, `t3.medium` instances.

I'd prefer to do something like this:
1. add a parameter `SpotInstancePrice` that can be any value ≥ 0 (`[0, +∞)`, or `[0, +∞[` in French/Italian notation).
2. add a condition `RequestSpotInstance` that checks whether `SpotInstancePrice > 0` — so we are requesting spot instances if and only if `SpotInstancePrice ≠ 0`.
3. Use the parameter and the condition in this template to set `amazonec2-request-spot-instance` to true or false depending on the condition's value, and `amazonec2-spot-price` to the parameter's value.

**Comment by benoitm974** [comment]
My initial request was more to see if you were open to this and willing to chnage. All your remarks are things we need for a generic approach I'll do them no issue, now that I see it won't stay unmerge :)

I'll resubmit in a couple of days.

**Comment by fquffio** [comment]
😂 Fair enough. Sorry again for the delay of my review. I saw the notification few days ago, but then I completely forgot about this.

File: runner.yml (hunk 4 above)
Comment (benoitm974) [reply runner.yml]: We have a disagreement here, I don't understanf why you liked those two. The MaxBuilds is not the number of concurrent builds BUT the number of build a runner can do before the manager will kill it... to avoid memory leaks, or mis-cache data etc ... putting it too low means you'll have your runner killed too much ...

Comment (benoitm974) [reply runner.yml]: I've added a separate parameter default to the same value as concurent builds

File: runner.yml (hunk 3 above)
Comment (benoitm974) [reply runner.yml]: Added a string field with a default value for this one.

File: runner.yml (hunk 5 above)
Comment (benoitm974) [reply runner.yml]: overlay2 is the default everywhere and recommanded by gitlab, yet docker machine does make it default yet ... I guess there is only good there.

for spot I've converted it to 2 parameter a Yes/No and a price one. at least there is no explaination on the >0 price enable spot etc ...

**Comment by benoitm974** [comment]
Any chance you can get those merge ? I'll move to you branch on my CI.
Thanks.
